Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+        protected bool _cursorMoved;
+
+        private TimeSpan _lastInvalidate;
+        private bool     _insertMode;
+        private bool     _suppressRedoStackReset;
+
+        private readonly UndoRedoStack _undoStack = new UndoRedoStack();
+        private readonly UndoRedoStack _redoStack = new UndoRedoStack();
+
+        private readonly Dictionary<Keys, KeyRepeatState> _keyRepeatStates;
+
+        public TextInputBase() {
+            _lastInvalidate  = DateTime.MinValue.TimeOfDay;
+            _keyRepeatStates = new Dictionary<Keys, KeyRepeatState>();
+
+            Input.Mouse.LeftMouseButtonReleased += OnGlobalMouseLeftMouseButtonReleased;
+            Input.Keyboard.KeyStateChanged      += OnGlobalKeyboardKeyStateChanged;
+        }
+
+        private void OnTextInput(string value) {
+            bool ctrlDown = this.IsCtrlDown;
+
+            foreach (char c in value) {
+                if (char.IsControl(c)) continue;
+                if (_font.GetCharacterRegion(c) == null) continue;
+
+                if (ctrlDown) {
+                    switch (c) {
+                        case 'c':
+                            HandleCopy();
+                            return;
+                        case 'x':
+                            HandleCut();
+                            return;
+                        case 'v':
+                            HandlePaste();
+                            return;
+                        case 'z':
+                            HandleUndo();
+                            return;
+                        case 'y':
+                            HandleRedo();
+                            return;
+                        case 'a':
+                            SelectAll();
+                            return;
+                    }
+                }
+
+                InputChar(c);
+            }
+        }
+
+        private void DeleteChars(int index, int length) {
+            if (length <= 0) return;
+
+            SetText(_text.Substring(0, index) + _text.Substring(index + length), true);
+        }
+
+        private string RemoveUnsupportedCharacters(string value) {
+            foreach (char c in value) {
+                if (_font.GetCharacterRegion(c) == null && !(_multiline && c == NEWLINE)) {
+                    return RemoveUnsupportedCharacters(value.Replace(c.ToString(), string.Empty));
+                }
+            }
+
+            return value;
+        }
+
+        private bool InsertChars(int index, string value, out int length) {
+            if (string.IsNullOrEmpty(value)) {
+                length = 0;
+                return false;
+            }
+
+            value = RemoveUnsupportedCharacters(value);
+
+            int startLength = _text.Length;
+
+            if (string.IsNullOrEmpty(_text)) {
+                SetText(value, true);
+            } else {
+                SetText(_text.Substring(0, index) + value + _text.Substring(index), true);
```

Comment:
There is https://docs.microsoft.com/en-us/dotnet/api/system.string.insert?view=netframework-4.8 for that

Reply:
Fixed!